profiler/internal/cmemprof/wrap_linux.c

```diff
@@ -0,0 +1,56 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022 Datadog, Inc.
+
+#define _GNU_SOURCE
+#include <dlfcn.h>
+#include <stdatomic.h>
+#include <stdint.h>
+
+#include "profiler_internal.h"
+
+/* The GNU linker supports a --wrap flag that lets wrap allocation calls
```

Comment:
Huge caveat here: wrapping `malloc`, etc. from within the program will *NOT* have any effect on memory allocation calls from dynamic libraries. This is true for using `--wrap` or `dlysm`-based wrapping. Only allocation calls from the statically-linked portions of the program will be wrapped. Wrapping allocation from dynamic libraries requires something like `LD_PRELOAD`-ing a shared library that defines a custom allocator.

Reply:
Update: pushed some code to find and edit the relocation tables at program startup to point allocations to our wrappers. Seems to work okay, added a test which calls `strdup` which is [documented](https://man7.org/linux/man-pages/man3/strdup.3.html#:~:text=The%20strdup()%20function%20returns,copies%20at%20most%20n%20bytes.) to allocate its result through `malloc`.